File: 1089.duplicate-zeros.java
/*
 * @lc app=leetcode id=1089 lang=java
 *
 * [1089] Duplicate Zeros
 */

// @lc code=start
class Solution {
    public void duplicateZeros(int[] arr) {
        int zeros = 0;
        for (int n : arr) {
            if (n == 0)
                zeros += 1;
        }
        int[] tmp = new int[arr.length + zeros];
        int idx = 0;
        for (int n : arr) {
            if (n == 0) {
                tmp[idx++] = n;
            }
            tmp[idx++] = n;
        }
        for (int ix = 0; ix != arr.length; ++ix) {
            arr[ix] = tmp[ix];
        }
    }
}
// @lc code=end

File: 1089.duplicate-zeros.cpp
/*
 * @lc app=leetcode id=1089 lang=cpp
 *
 * [1089] Duplicate Zeros
 */

// @lc code=start
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void duplicateZeros(vector<int>& arr) {
        auto num_zeros = count(arr.begin(), arr.end(), 0);
        vector<int> tmp(arr.size() + num_zeros, 0);
        auto iter = tmp.begin();
        for (const auto &el : arr) {
            if (el == 0) *(iter++) = 0;
            *(iter++) = el;
        }
        copy(tmp.begin(), tmp.begin() + arr.size(), arr.begin());
    }
};
// @lc code=end
